stf/src/sgx.rs

```diff
@@ -202,7 +201,10 @@ impl Stf {
         let call_hash = blake2_256(&call.encode());
         ext.execute_with(|| match call.call {
             TrustedCall::balance_set_balance(root, who, free_balance, reserved_balance) => {
-                Self::ensure_root(root)?;
+                if !validate_nonce(&root, call.nonce) {
```

Comment:
this is redundant code for every call, why not validate nonce before `match call.call`?
moreover, you may want to look at 
https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.or_else
You `if` can be simplified

Reply:
To move this within the match statement was necessary because the root account is only know after the matching.

Reply:
Thanks for the or_else hint - it's really much nicer. Fixed accordingly 👍